File: archive/pulumi-py/Pr7669/lib/s3_stack.py
"""
s3_stack.py

S3 buckets for API logs with lifecycle policies.
"""

import pulumi
import pulumi_aws as aws
from pulumi import ResourceOptions
from typing import Optional


class S3Stack(pulumi.ComponentResource):
    """S3 buckets for API logging."""

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        log_retention_days: int,
        tags: dict,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__("custom:s3:S3Stack", name, None, opts)

        # Get AWS account ID and region for unique naming
        caller_identity = aws.get_caller_identity()
        region = aws.get_region()

        # API logs bucket with globally unique name
        bucket_name = f"api-logs-{environment_suffix}-{caller_identity.account_id}-{region.name}"

        self.api_logs_bucket = aws.s3.BucketV2(
            f"api-logs-{environment_suffix}",
            bucket=bucket_name,
            tags={**tags, "Name": f"api-logs-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        # Enable versioning
        aws.s3.BucketVersioningV2(
            f"api-logs-versioning-{environment_suffix}",
            bucket=self.api_logs_bucket.id,
            versioning_configuration=aws.s3.BucketVersioningV2VersioningConfigurationArgs(
                status="Enabled",
            ),
            opts=ResourceOptions(parent=self)
        )

        # Block public access
        aws.s3.BucketPublicAccessBlock(
            f"api-logs-public-access-block-{environment_suffix}",
            bucket=self.api_logs_bucket.id,
            block_public_acls=True,
            block_public_policy=True,
            ignore_public_acls=True,
            restrict_public_buckets=True,
            opts=ResourceOptions(parent=self)
        )

        # Lifecycle policy
        aws.s3.BucketLifecycleConfigurationV2(
            f"api-logs-lifecycle-{environment_suffix}",
            bucket=self.api_logs_bucket.id,
            rules=[
                aws.s3.BucketLifecycleConfigurationV2RuleArgs(
                    id="expire-old-logs",
                    status="Enabled",
                    expiration=aws.s3.BucketLifecycleConfigurationV2RuleExpirationArgs(
                        days=log_retention_days,
                    ),
                    noncurrent_version_expiration=(
                        aws.s3.BucketLifecycleConfigurationV2RuleNoncurrentVersionExpirationArgs(
                            noncurrent_days=log_retention_days,
                        )
                    ),
                )
            ],
            opts=ResourceOptions(parent=self)
        )

        # Server-side encryption
        aws.s3.BucketServerSideEncryptionConfigurationV2(
            f"api-logs-encryption-{environment_suffix}",
            bucket=self.api_logs_bucket.id,
            rules=[
                aws.s3.BucketServerSideEncryptionConfigurationV2RuleArgs(
                    apply_server_side_encryption_by_default=(
                        aws.s3.BucketServerSideEncryptionConfigurationV2RuleApplyServerSideEncryptionByDefaultArgs(
                            sse_algorithm="AES256",
                        )
                    ),
                )
            ],
            opts=ResourceOptions(parent=self)
        )

        self.register_outputs({
            "api_logs_bucket_name": self.api_logs_bucket.bucket,
            "api_logs_bucket_arn": self.api_logs_bucket.arn,
        })

File: archive/analysis-py/Pr8476/lib/s3-stack.ts
/**
 * S3 Stack - Artifact storage with encryption and lifecycle rules
 *
 * This stack creates an S3 bucket for storing CodePipeline artifacts with:
 * - AES256 encryption at rest
 * - Versioning enabled
 * - Lifecycle rules to manage old artifacts
 * - Public access blocked
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface S3StackArgs {
  environmentSuffix: string;
  region: string;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class S3Stack extends pulumi.ComponentResource {
  public readonly artifactBucketName: pulumi.Output<string>;
  public readonly artifactBucketArn: pulumi.Output<string>;

  constructor(
    name: string,
    args: S3StackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:cicd:S3Stack', name, args, opts);

    // Artifact Bucket
    const artifactBucket = new aws.s3.Bucket(
      `pipeline-artifacts-${args.environmentSuffix}`,
      {
        bucket: `pipeline-artifacts-${args.environmentSuffix}`,
        tags: args.tags,
      },
      { parent: this }
    );

    // Enable Versioning
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const bucketVersioning = new aws.s3.BucketVersioningV2(
      `pipeline-artifacts-versioning-${args.environmentSuffix}`,
      {
        bucket: artifactBucket.id,
        versioningConfiguration: {
          status: 'Enabled',
        },
      },
      { parent: this }
    );

    // Server-Side Encryption
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const bucketEncryption =
      new aws.s3.BucketServerSideEncryptionConfigurationV2(
        `pipeline-artifacts-encryption-${args.environmentSuffix}`,
        {
          bucket: artifactBucket.id,
          rules: [
            {
              applyServerSideEncryptionByDefault: {
                sseAlgorithm: 'AES256',
              },
              bucketKeyEnabled: true,
            },
          ],
        },
        { parent: this }
      );

    // Lifecycle Rules
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const bucketLifecycle = new aws.s3.BucketLifecycleConfigurationV2(
      `pipeline-artifacts-lifecycle-${args.environmentSuffix}`,
      {
        bucket: artifactBucket.id,
        rules: [
          {
            id: 'expire-old-artifacts',
            status: 'Enabled',
            expiration: {
              days: 30,
            },
            noncurrentVersionExpiration: {
              noncurrentDays: 7,
            },
          },
          {
            id: 'transition-to-ia',
            status: 'Enabled',
            transitions: [
              {
                days: 30,
                storageClass: 'STANDARD_IA',
              },
            ],
          },
        ],
      },
      { parent: this }
    );

    // Block Public Access
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const bucketPublicAccessBlock = new aws.s3.BucketPublicAccessBlock(
      `pipeline-artifacts-public-access-block-${args.environmentSuffix}`,
      {
        bucket: artifactBucket.id,
        blockPublicAcls: true,
        blockPublicPolicy: true,
        ignorePublicAcls: true,
        restrictPublicBuckets: true,
      },
      { parent: this }
    );

    // Bucket Ownership Controls - enforce bucket owner for all objects
    // This is required for ACL-free bucket configuration (modern S3 best practice)
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const bucketOwnershipControls = new aws.s3.BucketOwnershipControls(
      `pipeline-artifacts-ownership-${args.environmentSuffix}`,
      {
        bucket: artifactBucket.id,
        rule: {
          objectOwnership: 'BucketOwnerEnforced',
        },
      },
      { parent: this }
    );

    this.artifactBucketName = artifactBucket.bucket;
    this.artifactBucketArn = artifactBucket.arn;

    this.registerOutputs({
      artifactBucketName: this.artifactBucketName,
      artifactBucketArn: this.artifactBucketArn,
    });
  }
}
